src/v/kafka/server/handlers/fetch.cc

```diff
@@ -147,7 +147,12 @@ static ss::future<read_result> read_from_partition(
     co_await std::move(rdr.reader).release()->finally();
 
     if (e) {
-        std::rethrow_exception(e);
+        vlog(
```

Comment:
You mention in the cover page that there are a couple of exceptions we want to retry on, but here we are changing the behavior for all exceptions, right?

What is the contract of this read_from_partitions method? It has the "3-state" problem, i.e., it returns a `ss::future<result<>>` so you can have a valid result, an error result or an exceptional future. Is there intended to be an important semantic difference between the last two cases?

Reply:
Or is it simply a bug that the lower layers allow the exceptional future case at all, instead of mapping to a result?

There was a similar thread on slack recently.

Reply:
Someone from the enterprise team may need to chime in here. My understanding is the following though.

Kafka clients seem to categorize errors as retriable or not. At least from looking at franz-go and librdkafka they don't seem to have any special handling for various error codes beyond this. If an error code is retriable the client library can retry it without informing the user of the library. Otherwise the client has to propagate the error up to the user and let them decide how to handle it.

`read_from_partitions` does return read_results with either of these error types. And I suppose dropping the client connection may be considered a non-retriable error? I'm not sure if it's really intended behavior though and it seems like we'd want to explicitly classify all of the exceptions being caught by `read_from_partition` as retriable or not within that method it's self. All of the exceptions being caught are from the `model::record_batch_reader` which either returns the data requested or throws an exception if an issue is encountered while reading.

Reply:
> And I suppose dropping the client connection may be considered a non-retriable error?

No, it is retry-able, pretty sure. Since that kind of thing can happen for any old reason, eg., a brief blip in connectivity and then the client needs to be robust to it.

Reply:
> All of the exceptions being caught are from the model::record_batch_reader which either returns the data requested or throws an exception if an issue is encountered while reading. So in that sense there isn't a bug in the lower layer allowing these exceptions through. Its entirely intended behavior.

The bug wouldn't be that `model::record_batch_reader` throws exceptions, but that whoever wrote `read_from_partition` decided to use `ss::future<result<>>` as the return type, so I would expect in that case one of the following would be true:

 A) All errors will be reported as a failed `result`, the future itself will never be exceptional. So in that case `read_from_partition` needs to remap exceptions to results in places that use exceptions.
 B) Errors may be reported either via exceptional future or failed result and there is some important semantic difference between those two cases.

I guess if it's (B) I would hope to discover what the difference is as it isn't documented here.

Reply:
So is this kafka storage error a retryable one?

Reply:
> So is this kafka storage error a retryable one?

Yeah, the Kafka storage error is retriable.

Reply:
> The bug wouldn't be that `model::record_batch_reader` throws exceptions, but that whoever wrote `read_from_partition` decided to use `ss::future<result<>>` as the return type, so I would expect in that case one of the following would be true:
> 
> A) All errors will be reported as a failed `result`, the future itself will never be exceptional. So in that case `read_from_partition` needs to remap exceptions to results in places that use exceptions.
> B) Errors may be reported either via exceptional future or failed result and there is some important semantic difference between those two cases.

The hope of this PR is to convince folks it should be (A). Right now it is (B) with exceptional futures causing the client connection to be dropped and failed results causing a Kafka error code to be returned for that particular partition.

Reply:
> > The bug wouldn't be that `model::record_batch_reader` throws exceptions, but that whoever wrote `read_from_partition` decided to use `ss::future<result<>>` as the return type, so I would expect in that case one of the following would be true:
> > A) All errors will be reported as a failed `result`, the future itself will never be exceptional. So in that case `read_from_partition` needs to remap exceptions to results in places that use exceptions.
> > B) Errors may be reported either via exceptional future or failed result and there is some important semantic difference between those two cases.
> 
> The hope of this PR is to convince folks it should be (A). Right now it is (B) with exceptional futures causing the client connection to be dropped and failed results causing a Kafka error code to be returned for that particular partition.

I took a look and I think the intention is (A), but when it comes to things like gate_closed_exception they just leak through and we aren't very disciplined about it.

What if we took a conservative approach and only converted to retry behavior for specific exceptions like gate_closed_exception?

Reply:
It's possible, but if dropping the connection is retriable anyway whats the difference between (A) and (B) other than losing the data read from other partitions in (B)? 

Maybe we can change the log level between known and unknown exceptions? I.e, the usual `info` log for  `gate_closed_exception` and `download_exception` then a `warn` log for any unknown exceptions?

Reply:
> It's possible, but if dropping the connection is retriable anyway whats the difference between (A) and (B) other than losing the data read from other partitions in (B)?

Right, probably nothing and it's fine. But because dropping the connection is like a big reset button on lots of state, its not immediately obvious that moving to retriable strategy doesn't implicit ignore something important happening in the big reset case when connection is dropped.

overall i think we should do this, and we have lots of testing to help us. being conservative i'd think now is a good time to merge before we start testing for next release, but i'd be less inclined to backport given this isn't tied to any incident or bug.

Reply:
@ballard26 WDYT?

Reply:
Sorry, missed this reply. Yeah, it makes a lot of sense. I've changed the PR to only avoid disconnecting on only the two exceptions I'm aware of. Both of which are easily retriable without any issues.